networkx/generators/joint_degree_seq.py

```diff
@@ -0,0 +1,286 @@
+       with a Target Joint Degree Matrix and Beyond", IEEE Infocom, 2015.
+
+    Examples
+    --------
+    >>> import networkx as nx
+    >>> joint_degree_dict = {1: {4: 1},
+    ...                      2: {2: 2, 3: 2, 4: 2},
+    ...                      3: {2: 2, 4: 1},
+    ...                      4: {1: 1, 2: 2, 3: 1}}
+    >>> G=nx.joint_degree_model(joint_degree_dict)
+    >>>
+
+    """
+    
+    if not is_valid_joint_degree(nkk):
+        msg = 'Input joint degree dict not realizable as a simple graph'
+        raise nx.NetworkXError(msg)
+
+    if not seed is None:
+        random.seed(seed)
+
+    # compute degree vector from joint degree nkk
+    nk = { deg:(sum(val for val in nkk[deg].itervalues())/deg) for deg in nkk if deg>0 }
+    
+    # start with empty N-node graph
+    N = int(sum(deg for deg in nk.itervalues()))
+    G = nx.empty_graph(N)
+    
+    
+    # for a given degree group, keep the list of all node ids 
+    h_degree_nodelist = {}
+    
+    # for a given node, keep track of the remaining stubs to be added. 
+    h_node_residual = {}
+    
+    # populate h_degree_nodelist and h_node_residual
+    nodeid = 0    
+    for degree, numNodes in nk.iteritems(): 
+        h_degree_nodelist[degree] = range(nodeid, nodeid+int(numNodes))         
+        for v in h_degree_nodelist[degree]:
+            h_node_residual[v] = degree  
+        nodeid += int(numNodes)
+                                        
+    # iterate over every degree pair (k,l) and add the number of edges given for each pair
+    for k in nkk:
+        for l in nkk[k]:
+            
+            # n_edges_add is the number of edges to add for the degree pair (k,l) 
+            n_edges_add = nkk[k][l] 
+            
+            if (n_edges_add>0) and (k>=l):                                                                 
+
+                # number of nodes with degree k and l
+                k_size = nk[k]
+                l_size = nk[l]    
+                
+                # k_nodes and l_nodes consist of all nodes of degree k and l respectivelyu                
+                k_nodes = h_degree_nodelist[k]
+                l_nodes = h_degree_nodelist[l]
+
+                # k_unsat and l_unsat consist of nodes of degree k and l that are unsaturated
+                # i.e. those nodes that have at least one available stub
+                k_unsat = set(v for v in k_nodes if h_node_residual[v]>0)                        
+                
+                if k!=l:
+                    l_unsat = set(w for w in l_nodes if h_node_residual[w]>0) 
+                else:
+                    l_unsat = k_unsat
+                    n_edges_add = nkk[k][l]/2
+                    
+                while n_edges_add > 0:
+                    
+                    # randomly pick nodes v and w that have degrees k and l
+                    v = k_nodes[ random.randint(0,k_size-1) ]
```

Comment:
Use `random.randrange(k_size)` here, it matches Python's ubiquitous `range` idiom more closely.